## Introduction

This notebook is to explore and cluter neighbourhoods of Toronto.
First at all, we will extract Toronto neighbourhood information from the Wikipedia page using BeautifulSoup library.
Then, we will explore each neighbourhood with Foursquare.
The next step is to apply the k-means algorithm to cluster the neighbourhoods.
And finally we we will visualize the map of the clustered neighbourhoods using the folium library.

## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1. <a href="#item1">Download and Explore Dataset</a>

2. <a href="#item2">Explore Neighborhoods in New York City</a>

3. <a href="#item3">Analyze Each Neighborhood</a>

4. <a href="#item4">Cluster Neighborhoods</a>

5. <a href="#item5">Examine Clusters</a>    
</font>
</div>

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [1]:
pip install beautifulsoup4 lxml html5lib

Note: you may need to restart the kernel to use updated packages.


In [2]:
from bs4 import BeautifulSoup
import pandas as pd 
import numpy as np
from sklearn.cluster import KMeans
#!conda install -c conda-forge folium=0.5.0 --yes 
import folium
#!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

<a id='item1'></a>

## 1. Download and Explore Dataset

In [3]:
soup = BeautifulSoup(open("./ListPostalCodesCanada.html"))

Here we want to extract the table of the list of postal codes of Canada, so we use the find feature and suppose that the table containing the is the first one of the html page.

In [4]:
table_tag = soup.find("table")

The column labels of the dataframe would be the header of the table.

In [5]:
labels = []
first_tr_tag = table_tag.tr
for th_tag in first_tr_tag.find_all("th"):
    labels.append(th_tag.string.rstrip())

We will add each element of three columns of a row respectively into the lists of postcode, borough and neighbourhood, and create the dataframe from these lists.

In [6]:
postCodeList = []
BoroughList = []
NeighbourhoodList = []
for tr_tag in first_tr_tag.find_next_siblings("tr"):
    postCodeTag = tr_tag.td
    postCodeList.append(postCodeTag.string)
    BoroughTag = postCodeTag.find_next_sibling("td")
    BoroughList.append(BoroughTag.string)
    NeighbourhoodTag = BoroughTag.find_next_sibling("td")
    for s in NeighbourhoodTag.stripped_strings:
        NeighbourhoodTag = s
    NeighbourhoodList.append(NeighbourhoodTag)

In [7]:
df = pd.DataFrame(list(zip(postCodeList, BoroughList, NeighbourhoodList)), columns =labels) 

In [8]:
df.rename(columns={'Postcode':'PostalCode'}, inplace=True)

In [9]:
df = df[df.Borough != "Not assigned"]

In [10]:
for index in df.index[df.Neighbourhood == "Not assigned"].tolist():
    df.at[index, 'Neighbourhood'] = df.at[index, 'Borough']

We suppose that if two rows have the same postcode, they have also the same borough. Then only the elements in the column of the 'Neighbourhood' need to be aggragated.

In [11]:
df['Neighbourhood'] = df.groupby(['PostalCode'])['Neighbourhood'].transform(lambda x: ','.join(x))
df = df.drop_duplicates()
df.reset_index(drop=True, inplace=True)

In [12]:
print(df.shape)
df.head(11)

(103, 3)


PostalCode           Borough                    Neighbourhood
0         M3A        North York                        Parkwoods
1         M4A        North York                 Victoria Village
2         M5A  Downtown Toronto         Harbourfront,Regent Park
3         M6A        North York  Lawrence Heights,Lawrence Manor
4         M7A      Queen's Park                     Queen's Park
5         M9A         Etobicoke                 Islington Avenue
6         M1B       Scarborough                    Rouge,Malvern
7         M3B        North York                  Don Mills North
8         M4B         East York   Woodbine Gardens,Parkview Hill
9         M5B  Downtown Toronto          Ryerson,Garden District
10        M6B        North York                        Glencairn

Then we will construct a dataframe with geo location of the neighbourhoods

In [13]:
df_location = pd.read_csv('./Geospatial_Coordinates.csv')

In [14]:
df_location.rename(columns={'Postal Code':'PostalCode'}, inplace=True)

In [15]:
df.set_index('PostalCode', inplace=True)
df_location.set_index('PostalCode', inplace=True)
df_toronto = pd.concat([df, df_location], axis=1, join='inner')
df_toronto.reset_index(inplace=True)

In [16]:
print(df_toronto.shape)
df_toronto.head(11)

(103, 5)


PostalCode           Borough                    Neighbourhood   Latitude  \
0         M3A        North York                        Parkwoods  43.753259   
1         M4A        North York                 Victoria Village  43.725882   
2         M5A  Downtown Toronto         Harbourfront,Regent Park  43.654260   
3         M6A        North York  Lawrence Heights,Lawrence Manor  43.718518   
4         M7A      Queen's Park                     Queen's Park  43.662301   
5         M9A         Etobicoke                 Islington Avenue  43.667856   
6         M1B       Scarborough                    Rouge,Malvern  43.806686   
7         M3B        North York                  Don Mills North  43.745906   
8         M4B         East York   Woodbine Gardens,Parkview Hill  43.706397   
9         M5B  Downtown Toronto          Ryerson,Garden District  43.657162   
10        M6B        North York                        Glencairn  43.709577   

    Longitude  
0  -79.329656  
1  -79.315572  
2  -79.360636  
3  -79.464763  
4  -79.389494  
5  -79.532242  
6  -79.194353  
7  -79.352188  
8  -79.309937  
9  -79.378937  
10 -79.445073

We can see that Toronto is a big city composed with at least 103 neighbourhoods, as all rows contain at least one neighbourhood.
In order to facilite the further analysis, we will consider each PostalCode as a neighbourhood.

In [17]:
df_toronto = df_toronto.drop('Neighbourhood', axis=1).rename(columns={'PostalCode':'Neighbourhood'}).reset_index(drop=True)

In [18]:
df_toronto.head()

Neighbourhood           Borough   Latitude  Longitude
0           M3A        North York  43.753259 -79.329656
1           M4A        North York  43.725882 -79.315572
2           M5A  Downtown Toronto  43.654260 -79.360636
3           M6A        North York  43.718518 -79.464763
4           M7A      Queen's Park  43.662301 -79.389494

<a id='item2'></a>

## 2. Explore Neighbourhoods in Toronto

Now, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

At first, we will create a function to find the nearby venues of the neighborhoods in Toronto.

In [19]:
# @hidden cell
CLIENT_ID = 'O1WFSGOCLWLBBL0GMS5N1DGFJZFO0UWQWDJAUP1D1LIFGN3Z' # your Foursquare ID
CLIENT_SECRET = 'B5NTAIO5UT41F1KBFWI3TEDAKEQYFDQMQGUC23RYVW4ZZRKU' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

In [20]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
        
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name,
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood',
                  'Latitude',
                  'Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Then, we can run the above function on each neighborhood and create a new dataframe.

In [21]:
LIMIT=100
toronto_venues = getNearbyVenues(names=df_toronto['Neighbourhood'],
                                 latitudes=df_toronto['Latitude'],
                                 longitudes=df_toronto['Longitude']
                                 )

Let's check the resulting dataframe.

In [22]:
print(toronto_venues.shape)
toronto_venues.head()

(2243, 7)


Neighbourhood   Latitude  Longitude                   Venue  Venue Latitude  \
0           M3A  43.753259 -79.329656         Brookbanks Park       43.751976   
1           M3A  43.753259 -79.329656                     KFC       43.754387   
2           M3A  43.753259 -79.329656           Variety Store       43.751974   
3           M3A  43.753259 -79.329656           GreenWin pool       43.756232   
4           M4A  43.725882 -79.315572  Victoria Village Arena       43.723481   

   Venue Longitude        Venue Category  
0       -79.332140                  Park  
1       -79.333021  Fast Food Restaurant  
2       -79.333114     Food & Drink Shop  
3       -79.333842                  Pool  
4       -79.315635          Hockey Arena

Let's check how many venues were returned for each neighborhood.

In [23]:
toronto_venues.groupby('Neighbourhood').count()

Latitude  Longitude  Venue  Venue Latitude  Venue Longitude  \
Neighbourhood                                                                
M1B                   1          1      1               1                1   
M1C                   1          1      1               1                1   
M1E                   8          8      8               8                8   
M1G                   4          4      4               4                4   
M1H                   7          7      7               7                7   
M1J                   2          2      2               2                2   
M1K                   4          4      4               4                4   
M1L                   9          9      9               9                9   
M1M                   2          2      2               2                2   
M1N                   4          4      4               4                4   
M1P                   6          6      6               6                6   
M1R                   7          7      7               7                7   
M1S                   4          4      4               4                4   
M1T                  10         10     10              10               10   
M1V                   3          3      3               3                3   
M1W                  13         13     13              13               13   
M2H                   6          6      6               6                6   
M2J                  66         66     66              66               66   
M2K                   4          4      4               4                4   
M2L                   2          2      2               2                2   
M2N                  34         34     34              34               34   
M2P                   3          3      3               3                3   
M2R                   5          5      5               5                5   
M3A                   4          4      4               4                4   
M3B                   4          4      4               4                4   
M3C                  21         21     21              21               21   
M3H                  19         19     19              19               19   
M3J                   5          5      5               5                5   
M3K                   3          3      3               3                3   
M3L                   4          4      4               4                4   
...                 ...        ...    ...             ...              ...   
M6B                   4          4      4               4                4   
M6C                   3          3      3               3                3   
M6E                   6          6      6               6                6   
M6G                  16         16     16              16               16   
M6H                  20         20     20              20               20   
M6J                  63         63     63              63               63   
M6K                  19         19     19              19               19   
M6L                   4          4      4               4                4   
M6M                   4          4      4               4                4   
M6N                   4          4      4               4                4   
M6P                  23         23     23              23               23   
M6R                  15         15     15              15               15   
M6S                  40         40     40              40               40   
M7A                  39         39     39              39               39   
M7R                  10         10     10              10               10   
M7Y                  17         17     17              17               17   
M8V                  15         15     15              15               15   
M8W                  10         10     10              10               10   
M8X                   2          2   

Let's find out how many unique categories can be curated from all the returned venue.

In [24]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 274 uniques categories.


<a id='item3'></a>

## 3. Analyze Each Neighbourhood

We will now analyze each neighbourhood of Toronto by checking the mean of the frequency of occurrence of each category for each neighbourhood.

In [25]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0           M1B                0.0             0.0                0.0   
1           M1C                0.0             0.0                0.0   
2           M1E                0.0             0.0                0.0   
3           M1G                0.0             0.0                0.0   
4           M1H                0.0             0.0                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  ...  Train Station  Vegetarian / Vegan Restaurant  \
0               0.0  ...            0.0                            0.0   
1               0.0  ...            0.0                            0.0   
2               0.0  ...            0.0                            0.0   
3               0.0  ...            0.0                            0.0   
4               0.0  ...            0.0                            0.0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0               0.0          0.0                    0.0              0.0   
1               0.0          0.0                    0.0              0.0   
2               0.0          0.0                    0.0              0.0   
3               0.0          0.0                    0.0              0.0   
4               0.0          0.0                    0.0              0.0   

   Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0       0.0          0.0            0.0          0.0  
1       0.0          0.0            0.0          0.0  
2       0.0          0.0            0.0          0.0  
3       0.0          0.0            0.0          0.0  
4       0.0          0.0            0.0          0.0  

[5 rows x 275 columns]

Let's write a function to sort the venues in descending order.

In [26]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Then we can use the function to create a dataframe to display the top 10 venues for each neighborhood.

In [27]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

In [28]:
neighborhoods_venues_sorted

Neighbourhood       1st Most Common Venue      2nd Most Common Venue  \
0            M1B        Fast Food Restaurant                  Drugstore   
1            M1C                         Bar        Dumpling Restaurant   
2            M1E              Medical Center             Breakfast Spot   
3            M1G                 Coffee Shop          Korean Restaurant   
4            M1H          Athletics & Sports        Fried Chicken Joint   
5            M1J           Convenience Store                 Playground   
6            M1K            Department Store                Bus Station   
7            M1L                      Bakery                   Bus Line   
8            M1M                       Motel        American Restaurant   
9            M1N       General Entertainment            College Stadium   
10           M1P           Indian Restaurant                  Pet Store   
11           M1R                 Auto Garage  Middle Eastern Restaurant   
12           M1S              Breakfast Spot                     Lounge   
13           M1T                 Pizza Place        Fried Chicken Joint   
14           M1V                        Park                 Playground   
15           M1W          Chinese Restaurant              Grocery Store   
16           M2H        Fast Food Restaurant                Golf Course   
17           M2J              Clothing Store                Coffee Shop   
18           M2K                        Bank                       Café   
19           M2L                   Cafeteria                       Park   
20           M2N                 Coffee Shop           Ramen Restaurant   
21           M2P                        Bank                       Park   
22           M2R                    Pharmacy              Grocery Store   
23           M3A                        Park       Fast Food Restaurant   
24           M3B        Gym / Fitness Center                       Café   
25           M3C            Asian Restaurant                        Gym   
26           M3H                 Coffee Shop             Sandwich Place   
27           M3J      Furniture / Home Store             Massage Studio   
28           M3K                        Park                    Airport   
29           M3L               Shopping Mall              Grocery Store   
..           ...                         ...                        ...   
70           M6B                         Pub        Japanese Restaurant   
71           M6C                       Field                      Trail   
72           M6E                        Park                   Pharmacy   
73           M6G                        Café              Grocery Store   
74           M6H                    Pharmacy                Supermarket   
75           M6J                         Bar           Asian Restaurant   
76           M6K                        Café             Breakfast Spot   
77           M6L                        Park             Massage Studio   
78           M6M                         Bar             Sandwich Place   
79           M6N                    Bus Line              Grocery Store   
80           M6P                         Bar                       Café   
81           M6R              Breakfast Spot                  Gift Shop   
82           M6S                        Café                Coffee Shop   
83           M7A                 Coffee Shop                        Gym   
84           M7R                       Hotel                Coffee Shop   
85           M7Y          Light Rail Station                Yoga Studio   
86           M8V                   Pet Store                     Bakery   
87           M8W                 Pizza Place                       Pool   
88           M8X                        Park                      River   
89           M8Y              Baseball Field                Yoga Studio   
90           M8Z               Grocery Store     Thrift / Vintage Store   
91           M9B            

<a id='item4'></a>

## 4. Cluster Neighbourhoods

Run k-means to cluster the neighborhood into 5 clusters.

In [29]:
kclusters = 5
toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

And add the clustering labels into the created neighbourhood dataframe.

In [30]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
toronto_merged = df_toronto
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood', how='inner')

In [31]:
toronto_merged.head()

Neighbourhood           Borough   Latitude  Longitude  Cluster Labels  \
0           M3A        North York  43.753259 -79.329656               0   
1           M4A        North York  43.725882 -79.315572               2   
2           M5A  Downtown Toronto  43.654260 -79.360636               2   
3           M6A        North York  43.718518 -79.464763               2   
4           M7A      Queen's Park  43.662301 -79.389494               2   

    1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                    Park  Fast Food Restaurant                  Pool   
1   Portuguese Restaurant           Pizza Place           Coffee Shop   
2             Coffee Shop                  Park                  Café   
3  Furniture / Home Store        Clothing Store     Accessories Store   
4             Coffee Shop                   Gym   Japanese Restaurant   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0     Food & Drink Shop           Yoga Studio      Doner Restaurant   
1          Hockey Arena            Donut Shop          Dessert Shop   
2                   Pub                Bakery               Theater   
3   Arts & Crafts Store             Gift Shop           Event Space   
4                 Diner           Yoga Studio         Smoothie Shop   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0    Dim Sum Restaurant                 Diner        Discount Store   
1    Dim Sum Restaurant                 Diner        Discount Store   
2        Breakfast Spot    Mexican Restaurant                   Spa   
3    Miscellaneous Shop           Coffee Shop              Boutique   
4    Seafood Restaurant        Sandwich Place          Burger Joint   

  10th Most Common Venue  
0                Dog Run  
1                Dog Run  
2             Beer Store  
3          Women's Store  
4          Burrito Place

Then we will visualize the resulting clusters.

We use here geopy library to get the latitude and longitude values of New York City.

In [32]:
address = 'Toronto'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


And we create a map of Toronto, using the folium library, with neighborhoods superimposed on top.

In [33]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<a id='item5'></a>

## 5. Examine Clusters

In [34]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough 1st Most Common Venue 2nd Most Common Venue  \
0    North York                  Park  Fast Food Restaurant   
6   Scarborough  Fast Food Restaurant             Drugstore   
21         York                  Park              Pharmacy   
40   North York                  Park               Airport   
66   North York                  Bank                  Park   
85  Scarborough                  Park            Playground   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Pool     Food & Drink Shop           Yoga Studio   
6     Dim Sum Restaurant                 Diner        Discount Store   
21         Women's Store  Fast Food Restaurant                Market   
40           Snack Place           Yoga Studio            Donut Shop   
66     Electronics Store           Yoga Studio    Dim Sum Restaurant   
85           Coffee Shop           Yoga Studio            Donut Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0       Doner Restaurant    Dim Sum Restaurant                 Diner   
6                Dog Run      Doner Restaurant            Donut Shop   
21            Donut Shop    Dim Sum Restaurant                 Diner   
40    Dim Sum Restaurant                 Diner        Discount Store   
66                 Diner        Discount Store               Dog Run   
85          Dessert Shop    Dim Sum Restaurant                 Diner   

   9th Most Common Venue 10th Most Common Venue  
0         Discount Store                Dog Run  
6    Dumpling Restaurant            College Gym  
21        Discount Store                Dog Run  
40               Dog Run       Doner Restaurant  
66      Doner Restaurant             Donut Shop  
85        Discount Store                Dog Run

In [35]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough 1st Most Common Venue 2nd Most Common Venue  \
62  Central Toronto                Garden            Donut Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
62          Dessert Shop    Dim Sum Restaurant                 Diner   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
62        Discount Store               Dog Run      Doner Restaurant   

   9th Most Common Venue 10th Most Common Venue  
62           Yoga Studio       Department Store

In [36]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough   1st Most Common Venue          2nd Most Common Venue  \
1          North York   Portuguese Restaurant                    Pizza Place   
2    Downtown Toronto             Coffee Shop                           Park   
3          North York  Furniture / Home Store                 Clothing Store   
4        Queen's Park             Coffee Shop                            Gym   
7          North York    Gym / Fitness Center                           Café   
8           East York    Fast Food Restaurant                    Pizza Place   
9    Downtown Toronto             Coffee Shop                 Clothing Store   
10         North York                     Pub            Japanese Restaurant   
11          Etobicoke                    Bank                    Golf Course   
12        Scarborough                     Bar            Dumpling Restaurant   
13         North York        Asian Restaurant                            Gym   
14          East York                Pharmacy               Asian Restaurant   
15   Downtown Toronto             Coffee Shop                          Hotel   
16               York                   Field                          Trail   
17          Etobicoke                Pharmacy                   Liquor Store   
18        Scarborough          Medical Center                 Breakfast Spot   
19       East Toronto       Health Food Store                    Music Venue   
20   Downtown Toronto             Coffee Shop                   Cocktail Bar   
22        Scarborough             Coffee Shop              Korean Restaurant   
23          East York     Sporting Goods Shop                    Coffee Shop   
24   Downtown Toronto             Coffee Shop                           Café   
25   Downtown Toronto                    Café                  Grocery Store   
26        Scarborough      Athletics & Sports            Fried Chicken Joint   
27         North York    Fast Food Restaurant                    Golf Course   
28         North York             Coffee Shop                 Sandwich Place   
29          East York       Indian Restaurant                 Sandwich Place   
30   Downtown Toronto             Coffee Shop                           Café   
31       West Toronto                Pharmacy                    Supermarket   
32        Scarborough       Convenience Store                     Playground   
33         North York          Clothing Store                    Coffee Shop   
..                ...                     ...                            ...   
68    Central Toronto                Bus Line               Sushi Restaurant   
69       West Toronto                     Bar                           Café   
70          Etobicoke             Pizza Place             Chinese Restaurant   
71        Scarborough             Auto Garage      Middle Eastern Restaurant   
72         North York                Pharmacy                  Grocery Store   
73    Central Toronto          Clothing Store                    Coffee Shop   
74    Central Toronto             Coffee Shop                 Sandwich Place   
75       West Toronto          Breakfast Spot                      Gift Shop   
76        Mississauga                   Hotel                    Coffee Shop   
77          Etobicoke             Pizza Place                           Park   
78        Scarborough          Breakfast Spot                         Lounge   
79    Central Toronto          Sandwich Place                    Pizza Place   
80   Downtown Toronto                    Café                            Bar   
81       West Toronto                    Café                    Coffee Shop   
82        Scarborough             Pizza Place            Fried Chicken Joint   
83    Central Toronto                   Trail                   Tennis Court   
84   Downtown Toronto                    Café  Vegetarian / Vegan Restaurant   
86    Central Toronto             Coffee Shop                            Pub   
87   Downtown 

In [37]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough 1st Most Common Venue 2nd Most Common Venue  \
35         East York                  Park     Convenience Store   
45        North York             Cafeteria                  Park   
64              York                  Park           Yoga Studio   
91  Downtown Toronto                  Park            Playground   
98         Etobicoke                  Park                 River   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
35         Metro Station    Dim Sum Restaurant                 Diner   
45            Donut Shop          Dessert Shop    Dim Sum Restaurant   
64             Drugstore    Dim Sum Restaurant                 Diner   
91                 Trail           Yoga Studio      Doner Restaurant   
98           Yoga Studio      Doner Restaurant          Dessert Shop   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
35        Discount Store               Dog Run      Doner Restaurant   
45                 Diner        Discount Store               Dog Run   
64        Discount Store               Dog Run      Doner Restaurant   
91      Department Store          Dessert Shop    Dim Sum Restaurant   
98    Dim Sum Restaurant                 Diner        Discount Store   

   9th Most Common Venue 10th Most Common Venue  
35            Donut Shop            Yoga Studio  
45      Doner Restaurant              Drugstore  
64            Donut Shop    Dumpling Restaurant  
91                 Diner         Discount Store  
98               Dog Run             Donut Shop

In [38]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough       1st Most Common Venue 2nd Most Common Venue  \
57   North York  Construction & Landscaping        Baseball Field   
101   Etobicoke              Baseball Field           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57            Yoga Studio   Dumpling Restaurant                 Diner   
101   Dumpling Restaurant                 Diner        Discount Store   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57         Discount Store               Dog Run      Doner Restaurant   
101               Dog Run      Doner Restaurant            Donut Shop   

    9th Most Common Venue       10th Most Common Venue  
57             Donut Shop                    Drugstore  
101             Drugstore  Eastern European Restaurant